demos/portal-embed-demo/public/js/ttt/audio-controller.js

```diff
@@ -0,0 +1,369 @@
+        // controller is not the point for this demo.
+        if(this.playlistIndex !== 0){
+            this.playlistIndex--;
+            this._createNewAudio(this.playlist[this.playlistIndex])
+            if(this._isPlaying()) {
+                // If the controller was playing an audio, 
+                // start playing the new audio
+                this.play();
+            }
+        }
+
+        this._updateController();
+    }
+
+    /**
+     * Hiding the controller with anmimation
+     */
+    hide() {
+        this.style.transition = 'bottom 0.6s';
+        this.style.bottom = '-80px';
+    }
+
+    /**
+     * Showing the controller with anmimation
+     */
+    show() {
+        this.style.transition = 'bottom 0.2s';
+        this.style.bottom = '0px';
+    }
+
+    /**
+     * Adding anything required to do on portal activate
+     */
+    handleActivation() {
+        if(this._isPlaying()) {
+            // change the left transition position
+            const currentLeft = this._getAxisCurrentLeft();
+            const left = this.timeline.offsetWidth;
+            const duration = this.durationSec * (1 - currentLeft/left);
+            this.axis.style.transition = `${duration}s linear`;
+            this.axis.style.left = `${left}px`;
+        }
+    }
+
+    /**
+     * Adding event listeners for the buttons
+     */
+    _hookEvents() {
+        this.playButton.addEventListener('click', evt => {
+            if(this._isPlaying()){
+                this.pause();
+                return;
+            }
+            this.play();
+        });
+        this.nextButton.addEventListener('click', evt => this.next());
+        this.prevButton.addEventListener('click', evt => this.prev());
+        this.audio.addEventListener("ended", evt => {
+            this.pause();
+            this._createNewAudio(this.playlist[this.playlistIndex]);
+       });
+    }
+
+    /**
+     * Creating a new audio element
+     * @param {String} id - audio ID
+     */
+    _createNewAudio(id) {
+
+        // Simply replace the exisiting audio element with a new one
+        if(this.root.querySelector('audio')){
+            this.root.removeChild(this.root.querySelector('audio'));
+        }
+        const audio = document.createElement('audio');
+        const source = document.createElement('source');
+
+        // We could build streaming API for this in the future
+        source.src = `/mp3/${id}.mp3`;
+        audio.appendChild(source);
+        this.root.appendChild(audio);
+        this.audio = audio;
+
+        // Resetting the progress bar
+        this.durationSec = audioList[id].durationSec;
+        this.audioTitle.innerHTML
+            = `${audioList[id].title} - Totally Tooling Tips`;
+        this.axis.style.transition = '';
+        this.axis.style.left = '0px';
+    }
+
+    /**
+     * Update the prev/next button status
+     */
+    _updateController() {
+        if(this._hasPrev()) {
```

Comment:
nit: Also, I wonder if we couldn't just define a "disabled" class in the CSS and make this implementation shorter:

```javascript
this.prevButton.classList.toggle('disabled', !this._hasPrev());
this.prevButton.classList.toggle('disabled', !this._hasNext());
```

Reply:
This works beautifully. I changed it to your recommendation and 1) removed the initial style of the button (which was set to opacity 0.3) 2) called _updateController() in the constructor and 3) removed any unnecessary operations around button opacity.

Reply:
(just realized I meant to write `nextButton` on the second line, but I assume you caught that mistake :wink:)